core/src/test/java/io/grpc/util/RoundRobinLoadBalancerTest.java

```diff
@@ -360,6 +377,25 @@ private Subchannel createMockSubchannel() {
     return subchannel;
   }
 
+  private void mockSubchannelAttributes(Subchannel subchannel, Attributes attributes) {
+    when(subchannel.getAttributes()).thenReturn(attributes);
+  }
+
+  private void mockSubchannelState(Subchannel subchannel, ConnectivityStateInfo state) {
+    mockSubchannelAttributes(subchannel, Attributes.newBuilder().set(STATE_INFO,
+        new AtomicReference<ConnectivityStateInfo>(state)).build());
+  }
+
+  private void mockSubchannelToConnect(final Subchannel subchannel) {
```

Comment:
As I mentioned in the comment above, it should be only LB who will update STATE_INFO. In this particular case, `requestConnection()` should trigger a `handleSubchannelState()` instead of updating the STATE_INFO directly.

Also, this stubbing logic is too far away from the test logic, which is not easy to follow. I would prefer explicitly verifying `requestConnection()` and call `handleSubchannelState()` from the test method. It seems the current call sites don't even care if `requestConnection()` is called, so you could skip the verification and just call `handleSubchannelState()` to trigger state change.